yt_dlp/extractor/eplus.py

```diff
@@ -29,14 +34,99 @@ class EplusIbIE(InfoExtractor):
             'No video formats found!',
             'Requested format is not available',
         ],
+    }, {
+        'url': 'https://live.eplus.jp/sample',
+        'info_dict': {
+            'id': 'stream1ng20210719-test-005',
+            'title': 'Online streaming test for DRM',
+            'live_status': 'was_live',
+            'release_date': '20210719',
+            'release_timestamp': 1626703200,
+            'description': None,
+        },
+        'params': {
+            'skip_download': True,
+            'ignore_no_formats_error': True,
+        },
+        'expected_warnings': [
+            'Could not find the playlist URL. This event may not be accessible',
+            'No video formats found!',
+            'Requested format is not available',
+            'This video is DRM protected',
+        ],
+    }, {
+        'url': 'https://live.eplus.jp/2053935',
+        'info_dict': {
+            'id': '331320-0001-001',
+            'title': '丘みどり2020配信LIVE Vol.2 ～秋麗～ 【Streaming+(配信チケット)】',
+            'live_status': 'was_live',
+            'release_date': '20200920',
+            'release_timestamp': 1600596000,
+        },
+        'params': {
+            'skip_download': True,
+            'ignore_no_formats_error': True,
+        },
+        'expected_warnings': [
+            'Could not find the playlist URL. This event may not be accessible',
+            'No video formats found!',
+            'Requested format is not available',
+        ],
     }]
 
+    _USER_AGENT = 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/119.0.0.0 Safari/537.36 Edg/119.0.0.0'
+    _TEST_EVENT_URL = 'https://live.eplus.jp/2053935'
+
+    def _perform_login(self, id, password):
+        urlh = self._request_webpage(
+            self._TEST_EVENT_URL, None, note='Getting auth status', errnote='Unable to get auth status')
+        if urlh.url.startswith(self._TEST_EVENT_URL):
+            # already logged in
+            return
+
+        cltft_token = urlh.headers.get('X-CLTFT-Token')
+        if not cltft_token:
+            raise ExtractorError('Unable to get X-CLTFT-Token', expected=False)
+        self._set_cookie('live.eplus.jp', 'X-CLTFT-Token', cltft_token)
+
```

Comment:
I thought we are checking for the `ci_session` cookie to determine if the user is already logged in? The `ci_session` cookie needs to be set before a login attempt as well?

Reply:
> The `ci_session` cookie needs to be set before a login attempt as well?

Yes. The flow is as follows:

- Access `https://live.eplus.jp/[a-zA-Z0-9_-]+`
- Get `ci_session` , 302 redirect to `https://live.eplus.jp/member/auth?q=<base64>`
- Get other cookies, 302 redirect to `https://live.eplus.jp/member/auth?op=init`
- POST id and password with cookies to `https://live.eplus.jp/member/api/v1/FTAuth/idpw`
- POST the above login info and other required data to `https://live.eplus.jp/member/auth`
- 302 redirect to `https://live.eplus.jp/[a-zA-Z0-9_-]+?q=<base64>`
- 302 redirect to `https://live.eplus.jp/[a-zA-Z0-9_-]+`

Reply:
Ah, gotcha.

> The most suitable URL is the one input by the user because it should be valid. 

Agreed

> Getting the input URL in `_perform_login()` , looks not so good...

I think for this extractor we shouldn't redefine `_perform_login()` and instead call our own `_login()` method when needed. What do you think of this: https://github.com/bashonly/yt-dlp/commit/bb84d2e7a53ddc4c8368439de01c9d0e0af6c76c

Reply:
Nice `self._get_login_info()`! I didn't know that.

And we need get the webpage again after login:

```diff
              self._login(username, password, urlh)
+             webpage = self._download_webpage(
+                 url, video_id, headers={'User-Agent': self._USER_AGENT})

```